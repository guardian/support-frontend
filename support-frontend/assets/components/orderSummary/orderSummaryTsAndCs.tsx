import { css } from '@emotion/react';
import {
	from,
	neutral,
	palette,
	space,
	textSans14,
} from '@guardian/source/foundations';
import { type ContributionType } from 'helpers/contributions';
import type { CountryGroupId } from 'helpers/internationalisation/countryGroup';
import { productLegal } from 'helpers/legalCopy';
import type { ActiveProductKey } from 'helpers/productCatalog';
import type { Promotion } from 'helpers/productPrice/promotions';

const containerSummaryTsCs = css`
	border-radius: ${space[2]}px;
	background-color: ${neutral[97]};
	padding: ${space[3]}px;
`;
const productStartDate = css`
	display: block;
	${textSans14};
	color: #606060;
	background-color: ${palette.neutral[97]};
	border-radius: ${space[3]}px;
	padding: ${space[3]}px;
	margin-top: ${space[2]}px;
	${from.desktop} {
		margin-top: ${space[4]}px;
	}
	li + li {
		margin-top: ${space[2]}px;
	}
`;

interface OrderSummaryStartDateProps {
	startDate: string;
	productKey: ActiveProductKey;
}
export function OrderSummaryStartDate({
	startDate,
	productKey,
}: OrderSummaryStartDateProps): JSX.Element | null {
	const validProduct = [
		'GuardianWeeklyDomestic',
		'GuardianWeeklyRestOfWorld',
		'TierThree',
	].includes(productKey);
	if (validProduct) {
		console.log('*** productKey', productKey);
		const digitalBenefit =
			productKey === 'TierThree' ? (
				<li>Your digital benefits will start today.</li>
			) : null;
		const guardianWeeklyBenefit =
			productKey === 'TierThree' ||
			productKey === 'GuardianWeeklyDomestic' ||
			productKey === 'GuardianWeeklyRestOfWorld' ? (
				<li>
					Your Guardian Weekly subscription will start on {startDate}. Please
					allow 1 to 7 days after your start date for your magazine to arrive,
					depending on national post services.
				</li>
			) : null;
		return (
			<ul css={productStartDate}>
				{digitalBenefit}
				{guardianWeeklyBenefit}
			</ul>
		);
	}
	return null;
}

export interface OrderSummaryTsAndCsProps {
	productKey: ActiveProductKey;
	contributionType: ContributionType;
	countryGroupId: CountryGroupId;
	promotion?: Promotion;
	thresholdAmount?: number;
}
export function OrderSummaryTsAndCs({
	productKey,
	contributionType,
	countryGroupId,
	promotion,
	thresholdAmount = 0,
}: OrderSummaryTsAndCsProps): JSX.Element | null {
	// Proceeds with RegularContributionType only
	if (contributionType === 'ONE_OFF') {
		return null;
	}
	const period = contributionType === 'MONTHLY' ? 'month' : 'year';
	const tierThreeSupporterPlusTsAndCs = (
		<div css={containerSummaryTsCs}>
			{promotion && (
				<p>
					Youâ€™ll pay{' '}
					{productLegal(
						countryGroupId,
						contributionType,
						'/',
						thresholdAmount,
						promotion,
					)}{' '}
					afterwards unless you cancel. Offer only available to new subscribers
					who do not have an existing subscription with the Guardian.
				</p>
			)}
			{productKey === 'SupporterPlus' && (
				<>
					<p>Auto renews every {period} until you cancel.</p>
					<p>
						Cancel or change your support anytime. If you cancel within the
						first 14 days, you will receive a full refund.
					</p>
				</>
			)}
			{productKey === 'TierThree' && (
				<p>Auto renews every {period}. Cancel anytime.</p>
			)}
		</div>
	);
	const defaultOrderSummaryTsAndCs = (
		<div css={containerSummaryTsCs}>
			<p>Auto renews every {period} until you cancel.</p>
			<p>
				{['Contribution', 'OneTimeContribution'].includes(productKey)
					? 'Cancel or change your support anytime.'
					: 'Cancel anytime.'}
			</p>
		</div>
	);
	const orderSummaryTsAndCs: Partial<Record<ActiveProductKey, JSX.Element>> = {
		SupporterPlus: tierThreeSupporterPlusTsAndCs,
		TierThree: tierThreeSupporterPlusTsAndCs,
	};
	return orderSummaryTsAndCs[productKey] ?? defaultOrderSummaryTsAndCs;
}
