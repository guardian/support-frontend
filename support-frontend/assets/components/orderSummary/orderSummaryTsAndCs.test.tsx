import { render } from '@testing-library/react';
import type { CountryGroupId } from 'helpers/internationalisation/countryGroup';
import type {
	ActiveProductKey,
	ActiveRatePlanKey,
} from 'helpers/productCatalog';
import type { Promotion } from 'helpers/productPrice/promotions';
import { OrderSummaryTsAndCs } from './orderSummaryTsAndCs';

describe('orderSummaryTs&Cs Snapshot comparison', () => {
	const promotionTierThreeUnitedStatesMonthly: Promotion = {
		name: '$8 off for 12 months',
		description: 'Tier Three United States Monthly',
		promoCode: 'TIER_THREE_USA_MONTHLY',
		numberOfDiscountedPeriods: 12,
		discountedPrice: 37,
	};

	type OrderSummaryTestParams = [
		ActiveProductKey,
		CountryGroupId,
		ActiveRatePlanKey,
		number,
	];

	const orderSummaryProductKeys: OrderSummaryTestParams[] = [
		['GuardianAdLite', 'GBPCountries', 'Monthly', 0],
		['Contribution', 'AUDCountries', 'Annual', 0],
		['SupporterPlus', 'GBPCountries', 'Monthly', 12],
		['TierThree', 'UnitedStates', 'RestOfWorldMonthly', 45],
	];
	it.each(orderSummaryProductKeys)(
		`orderSummaryTs&Cs render product %s for region %s correctly`,
		(paymentProductKey, countryGroupId, activeRatePlanKey, amount) => {
			const promo: Promotion | undefined =
				paymentProductKey === 'TierThree' &&
				activeRatePlanKey === 'RestOfWorldMonthly' &&
				countryGroupId === 'UnitedStates'
					? promotionTierThreeUnitedStatesMonthly
					: undefined;
			const { container } = render(
				<OrderSummaryTsAndCs
					productKey={paymentProductKey}
					ratePlanKey={activeRatePlanKey}
					countryGroupId={countryGroupId}
					thresholdAmount={amount}
					promotion={promo}
				/>,
			);
			expect(container.textContent).toMatchSnapshot();
		},
	);
});
