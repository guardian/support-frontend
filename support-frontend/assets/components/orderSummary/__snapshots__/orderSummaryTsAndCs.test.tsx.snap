// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`orderSummaryTs&Cs Snapshot comparison orderSummaryTs&Cs render product Contribution for region AUDCountries correctly 1`] = `"Auto renews every year until you cancel.Cancel or change your support anytime."`;

exports[`orderSummaryTs&Cs Snapshot comparison orderSummaryTs&Cs render product GuardianAdLite for region GBPCountries correctly 1`] = `"Auto renews every month until you cancel.Cancel anytime."`;

exports[`orderSummaryTs&Cs Snapshot comparison orderSummaryTs&Cs render product SupporterPlus for region GBPCountries correctly 1`] = `"Auto renews every month until you cancel.Cancel or change your support anytime. If you cancel within the first 14 days, you will receive a full refund."`;

exports[`orderSummaryTs&Cs Snapshot comparison orderSummaryTs&Cs render product TierThree for region UnitedStates correctly 1`] = `"Youâ€™ll pay $37/month for the first 12 months, then $45/month afterwards unless you cancel. Offer only available to new subscribers who do not have an existing subscription with the Guardian.Auto renews every month. Cancel anytime."`;
